internal/secrets/encryptor_test.go

```
@@ -0,0 +1,285 @@
+	}
+}
+
+// Test that different strings EncryptBytes to different outputs
+func TestDifferentOutputs(t *testing.T) {
+	key, _ := generateRandomAESKey()
+	e := newEncryptor(key, nil)
+	messages := []string{
+		"This may or may",
+		"This is not the same as that",
+		"The end of that",
+		"Plants and animals",
+		"Snow, igloos, sunshine, unicords",
+	}
+
+	var crypts []string
+	for _, m := range messages {
+		encrypted, _ := e.EncryptBytes([]byte(m))
+		crypts = append(crypts, string(encrypted))
+	}
+
+	for _, c := range crypts {
+		if !isInSliceOnce(c, crypts) {
+			t.Fatalf("Duplicate encryption string: %v.", c)
+		}
+	}
+}
+
+func isInSliceOnce(item string, slice []string) bool {
+	found := 0
+	for _, s := range slice {
+		if item == s {
+			found++
+		}
+	}
+
+	return found == 1
+}
+
+func TestSampleNoRepeats(t *testing.T) {
+	key, _ := generateRandomAESKey()
+	e := newEncryptor(key, nil)
+
+	var crypts []string
+	for i := 0; i < 10000; i++ {
+		encrypted, _ := e.EncryptBytes(messageToEncrypt)
+		crypts = append(crypts, string(encrypted))
+	}
+
+	for _, item := range crypts {
+		if isInSliceOnce(item, crypts) == false {
+			t.Fatalf("Duplicate encrypted string found.")
+		}
+	}
+}
+
+func TestKeyMigration(t *testing.T) {
+	keyA, err := generateRandomAESKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	keyB, err := generateRandomAESKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	encryptorA := newEncryptor(keyA, nil)
+
+	message := "encrypted with Key A"
+	encryptedMessage, err := encryptorA.EncryptBytes([]byte(message))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// now rotate keys to use Key B
+	encryptorB := newEncryptor(keyB, keyA)
+	decryptedMessage, err := encryptorB.DecryptBytes(encryptedMessage)
+	if err != nil {
+		t.Fatalf("unable to Decrypt string: %v", err)
+	}
+
+	if string(decryptedMessage) != message {
+		t.Fatalf("messages do not match")
+	}
+}
+
+func TestEncryptAndDecryptBytesIfPossible(t *testing.T) {
+	initialKey, _ := generateRandomAESKey()
```

Check the error or add a code comment to explain why it is OK to ignore the error here in particular.